File: dcp-solutions/2024-11-03/solution.php
```php
class URLShortener {
    private $urlMap = array();
    private $alphabet = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';

    public function shorten($url) {
        $shortenedUrl = substr(str_shuffle($this->alphabet), 0, 6);
        if (isset($this->urlMap[$shortenedUrl])) {
            return $this->urlMap[$shortenedUrl];
        } else {
            $this->urlMap[$shortenedUrl] = $url;
            return $shortenedUrl;
        }
    }

    public function restore($short) {
        if (!isset($this->urlMap[$short])) {
            return null;
        }
        return $this->urlMap[$short];
    }
}
```

File: dcp-solutions/2024-11-03/solution.c
```c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <iostream>
#include <unordered_map>
using namespace std;

unordered_map<string, string> url_map;
int counter = 0;

string shorten(string url) {
    while (url_map.size() <= counter + 100)
        url_map[""] = "";

    char code[7];
    sprintf(code, "%d", counter++);
    for (; !isalnum(code[0]); counter++)
        sprintf(code, "%d", counter);
    for (; !isalnum(code[counter - 1]); counter++)
        sprintf(code, "%d", counter);

    return string(code) + url;
}

string restore(string code) {
    auto it = url_map.find(code.substr(0, code.size() - 1));
    if (it != url_map.end())
        return code + it->second;
    else
        return "";
}
```